enhancements/storage/csi-inline-vol-security.md

```diff
@@ -0,0 +1,343 @@
+effective because most clusters do not have a large number of `CSIDriver`
+objects, and these objects do not change frequently.
+
+The default behavior of this plugin will have an impact on user experience
+for any customers using inline volumes, in that the cluster administrator will
+have to decide on a policy for each CSI driver that they intend to use for
+inline volumes in unprivileged namespaces. This is a tradeoff between security
+and usability, and the default behavior favors a secure-by-default configuration
+over a more permissive configuration. This deserves some scrutiny during
+enhancement review, and approval from the product experience team.
+
+### Drawbacks
+
+See [Risks and Mitigations](#risks-and-mitigations).
+
+## Design Details
+
+### Open Questions [optional]
+
+N/A
+
+### Test Plan
+
+- Unit testing in tree, applied in a carry patch for maintainability.
+- Run e2e suite of the Shared Resource CSI driver with the plugin enabled.
+- End to end testing in openshift/origin, tested against the following:
+  - Build pods that utilize the Shared Resource CSI driver
+  - Workloads that utilize a `privileged` CSI driver in a `restricted` and
+    `baseline` namespace.
+
+### Graduation Criteria
+
+| OpenShift | Maturity |
+| --------- | -------- |
+| 4.12 | Tech Preview |
+| 4.13 | GA |
+
+#### Dev Preview -> Tech Preview
+
+- Admission plugin added to our fork of kube-apiserver
+- Test coverage in openshift/origin for pods with CSI volumes
+- Documentation for the admission plugin - use and configuration
+- Verify use with the Shared Resource CSI Driver
+
+#### Tech Preview -> GA
+
+- Seamless upgrade experience for the plugin, enabled by default.
+- Pod creation scale testing within acceptable limits.
+- End to end testing with pods using volumes from the Shared Resource CSI driver
+- Sufficient time for feedback
+- User facing documentation created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+#### Removing a deprecated feature
+
+N/A
+
+### Upgrade / Downgrade Strategy
+
+CSI drivers that are managed by OCP and support inline volumes will automatically
+have a default CSI ephemeral volume profile applied to the `CSIDriver` object by
+Cluster Storage Operator.
+
+Users of inline volumes who upgrade to an OCP release with the admission plugin
+enabled will need to add a CSI ephemeral volume profile for any unmanaged `CSIDriver`
+objects that should allow the use of inline volumes in unprivileged namespaces.
+
+### Version Skew Strategy
+
+N/A
+
+### Operational Aspects of API Extensions
+
+**TODO**
+
+[Shared Resource Validation](https://github.com/openshift/enhancements/blob/master/enhancements/cluster-scope-secret-volumes/shared-resource-validation.md#operational-aspects-of-api-extensions) can serve as an example here.
+
+#### New SLIs for the new admission validations to help cluster administrators and support
+
+#### Impact of existing API Server SLIs
```

Comment:
Thoughts here:

- Latency measures for pod APIs (or general kube apiserver latency measures)?
- Memory consumption of kube apiserver?

Reply:
Thanks, added some notes on this, there are a few existing metrics we can use for this.